torchtext/csrc/gpt2_bpe_tokenizer.cpp

```diff
@@ -397,18 +397,45 @@ std::vector<int64_t> GPT2BPEEncoder::Encode(const std::string& text) {
 
 std::string GPT2BPEEncoder::Decode(const std::vector<int64_t>& tokens) {
   std::string text;
+  std::vector<bool> special_token_flags(tokens.size());
   // setup converter for converting wide chars to/from chars
   using convert_type = std::codecvt_utf8<wchar_t>;
   std::wstring_convert<convert_type, wchar_t> converter;
 
-  for (const auto token : tokens) {
-    // get unicode string for given integer key
-    const std::string str = bpe_decoder_.at(token);
-    const std::wstring ws = converter.from_bytes(str);
-    for (wchar_t wchr : ws) {
-      // get output character from byte decoder for each wide character
-      unsigned char uchr = byte_decoder_.at(converter.to_bytes(wchr));
-      text.push_back(uchr);
+  for (int tok_idx = 0; tok_idx < tokens.size(); tok_idx++) {
+    const auto token = tokens[tok_idx];
+    std::string decoded_token;
+
+    if (added_tokens_decoder_.contains(token)) {
+      // string is a special token from extended vocab
+      decoded_token = added_tokens_decoder_.at(token);
+      special_token_flags[tok_idx] = true;
+    } else {
+      const std::string str = bpe_decoder_.at(token);
+      if (bpe_never_split_set_.find(str) != bpe_never_split_set_.end()) {
+        // string is a special token from known vocab
+        decoded_token = str;
+        special_token_flags[tok_idx] = true;
+      } else {
+        // string is a regular token from known vocab
+        const std::wstring ws = converter.from_bytes(str);
+        for (wchar_t wchr : ws) {
+          // get output character from byte decoder for each wide character
+          unsigned char uchr = byte_decoder_.at(converter.to_bytes(wchr));
+          decoded_token.push_back(uchr);
+        }
+      }
+    }
+
+    // fix left space(s) for special tokens
+    if (special_token_flags[tok_idx] == true &&
+        (tok_idx > 0 && special_token_flags[tok_idx - 1] == false)) {
```

Comment:
Just want to double check, does `&& special_token_flags[tok_idx - 1] == false` only ensure that we append a space if the token before the special token is a regular token? If so, why do we do this?

Reply:
It ensures that we insert space only if the last token wasn't special, because if the last token was special, it would have added space after it (through code at line 438). So we avoid having one extra space in case when two special tokens are next to each other.